File: visitor_pattern/visitor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
ThisとThatの２つのElementクラスがあり、それぞれにacceptメソッドがある。
Visitorクラスは、訪問する要素の型に基づいて適切なvisitメソッドを呼び出す。
UpVisitorとDownVisitorは、それぞれ異なる操作をThisとThatに対して行う。

Pythonの例ではgetattr関数とメソッド名の動的な生成を利用して、C++のオーバー
ロードを模倣している。
これによいｒ、新しいElementクラスを追加するときに、対応するvisitメソッドを
Visitorクラスに追加するだけで済む。
"""

class Element:
    def accept(self, visitor):
        visitor.visit(self)

class This(Element):
    def thiss(self):
        return "This"

class That(Element):
    def that(self):
        return "That"

class Visitor:
    def visit(self, element):
        method_name = 'visit_' + type(element).__name__
        method = getattr(self, method_name, self.visit_default)
        method(element)

    def visit_default(self, element):
        raise NotImplementedError("No visit_{} method".format(type(element).__name__))


class UpVisitor(Visitor):
    def visit_This(self, element):
        print("do Up on " + element.thiss())

    def visit_That(self, element):
        print("do Up on " + element.that())

class DownVisitor(Visitor):
    def visit_This(self, element):
        print("do Down on " + element.thiss())

    def visit_That(self, element):
        print("do Down on " + element.that())

aThis = This()
aThat = That()

up = UpVisitor()
down = DownVisitor()

aThis.accept(up)
aThat.accept(up)

aThis.accept(down)
aThat.accept(down)

File: visitor_pattern/visitor.cpp
#include <iostream>

/**
 * ビジターパターンとは、オブジェクトの操作をそのクラスから分離することで、
 * 新しい操作を追加するのを容易にするデザインパターンです。
 */

/// "Element" interface
class Element {
public:
    virtual ~Element() {}
    virtual void accept(class Visitor &v) = 0;
};

/// "ConcreteElement" classes
class This : public Element {
public:
    void accept(Visitor &v);
    std::string thiss() { return "This"; }
};

class That : public Element {
public:
    void accept(Visitor &v);
    std::string that() { return "That"; }
};


/// "Visitor" interface
class Visitor {
public:
    virtual ~Visitor() {}
    virtual void visit(This *e) = 0;
    virtual void visit(That *e) = 0;
};

/// "concreteVisitor" classes
class UpVisitor : public Visitor {
public:
    void visit(This *e) {
        std::cout << "do Up on " + e->thiss() << '\n';
    }
    void visit(That *e) {
        std::cout << "do Up on " + e->that() << '\n';
    }
};

class DownVisitor : public Visitor {
public:
    void visit(This *e) {
        std::cout << "do Down on " + e->thiss() << '\n';
    }
    void visit(That *e) {
        std::cout << "do Down on " + e->that() << '\n';
    }
};

/// Implement "accept" functions
void This::accept(Visitor &v) { v.visit(this); }
void That::accept(Visitor &v) { v.visit(this); }

int main() {
    This aThis;
    That aThat;

    UpVisitor up;
    DownVisitor down;

    aThis.accept(up);
    aThat.accept(up);

    aThis.accept(down);
    aThat.accept(down);
}

